## Import pandas & numpy

In [1]:
import pandas as pd
import numpy as np

## Import csv file

In [2]:
sharks = pd.read_csv("../Data_cleaning_manipulation/GSAF5.csv", encoding = 'cp1252')

## Preview the data (first 5 rows)

In [94]:
sharks.head()

case_number       date  year incident_type    country      area  \
0  2016.09.18.c  18-Sep-16  2016    Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016    Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016    Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   

                           location activity             name sex  ...  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  ...   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  ...   
2  New Smyrna Beach, Volusia County  Surfing             male   M  ...   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  ...   
4                       Bells Beach  Surfing             male   M  ...   

     species       investigator_or_source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.16   

  case_number.2 original_order    year_categories year_categories_decades  
0  2016.09.18.c           5993  from 2000 onwards       from 2010 onwards  
1  2016.09.18.b           5992  from 2000 onwards       from 2010 onwards  
2  2016.09.18.a           5991  from 2000 onwards       from 2010 onwards  
3    2016.09.17           5990  from 2000 onwards       from 2010 onwards  
4    2016.09.15           5989  from 2000 onwards       from 2010 onwards  

[5 rows x 24 columns]

## Make a copy of file

Before starting to work on the file, it is good practice to make a copy, so that any mistakes do not overwrite the original file

In [4]:
sharks_copy = sharks.copy()

## Start to look at the data

In [5]:
sharks_copy.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

We can see that a lot of these are objects when they shouldn't be (e.g. time or date)

In [6]:
sharks_copy.dtypes

Case Number               object
Date                      object
Year                       int64
Type                      object
Country                   object
Area                      object
Location                  object
Activity                  object
Name                      object
Sex                       object
Age                       object
Injury                    object
Fatal (Y/N)               object
Time                      object
Species                   object
Investigator or Source    object
pdf                       object
href formula              object
href                      object
Case Number.1             object
Case Number.2             object
original order             int64
Unnamed: 22               object
Unnamed: 23               object
dtype: object

In [7]:
#to find number of rows & columns. Output: (5992, 24)
sharks_copy.shape

(5992, 24)

In [8]:
sharks_copy['Sex '].value_counts()

M      4835
F       585
M         2
lli       1
.         1
N         1
Name: Sex , dtype: int64

In [9]:
sharks_copy['Country'].value_counts()

USA                       2116
AUSTRALIA                 1279
SOUTH AFRICA               565
PAPUA NEW GUINEA           133
NEW ZEALAND                125
                          ... 
WESTERN SAMOA                1
RED SEA / INDIAN OCEAN       1
KOREA                        1
EGYPT / ISRAEL               1
RED SEA                      1
Name: Country, Length: 203, dtype: int64

In [10]:
sharks_copy['Fatal (Y/N)'].value_counts()

N          4315
Y          1552
UNKNOWN      94
 N            8
N             1
n             1
#VALUE!       1
F             1
Name: Fatal (Y/N), dtype: int64

In [11]:
sharks_copy['Year'].value_counts().sort_values

<bound method Series.sort_values of 2015    139
2011    128
2014    125
0       124
2013    122
       ... 
1742      1
1758      1
1818      1
1822      1
1595      1
Name: Year, Length: 232, dtype: int64>

## Fix column headings

The column names look a bit difficult to work with, so I decided to improve them. 

First, I stripped the whitespaces in the headings.

In [12]:
sharks_copy.columns = sharks_copy.columns.str.strip()

Then I  set column names to lower case to make easier to work with

In [13]:
sharks_copy.columns = sharks_copy.columns.str.lower()

Then I replaced spaces in column names with underscores to make easier to work with

In [14]:
sharks_copy.columns = sharks_copy.columns.str.replace(' ', '_')

In [15]:
sharks_copy.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

## Rename 'type' column 

This column name is unclear, so I renamed it to incident_type

In [16]:
sharks_copy = sharks_copy.rename(columns={'type': 'incident_type'})

## Check if columns can be deleted

I noticed that a lot of values in two columns were nulls, so I checked how many missing values they had.

In [17]:
sharks_copy['unnamed:_22'].value_counts()

stopped here    1
Name: unnamed:_22, dtype: int64

In [18]:
sharks_copy['unnamed:_23'].value_counts()

change filename    1
Teramo             1
Name: unnamed:_23, dtype: int64

I also checked how many missing values there were for all columns

In [19]:
missing_values = sharks_copy.isna().sum()

In [20]:
missing_values

case_number                  0
date                         0
year                         0
incident_type                0
country                     43
area                       402
location                   496
activity                   527
name                       200
sex                        567
age                       2681
injury                      27
fatal_(y/n)                 19
time                      3213
species                   2934
investigator_or_source      15
pdf                          0
href_formula                 1
href                         3
case_number.1                0
case_number.2                0
original_order               0
unnamed:_22               5991
unnamed:_23               5990
dtype: int64

And checked the percentage of missing values per column, to get a quick overview

In [21]:
missing_values_perc = sharks_copy.isna().sum()/len(sharks_copy)

In [22]:
missing_values_perc

case_number               0.000000
date                      0.000000
year                      0.000000
incident_type             0.000000
country                   0.007176
area                      0.067089
location                  0.082777
activity                  0.087951
name                      0.033378
sex                       0.094626
age                       0.447430
injury                    0.004506
fatal_(y/n)               0.003171
time                      0.536215
species                   0.489653
investigator_or_source    0.002503
pdf                       0.000000
href_formula              0.000167
href                      0.000501
case_number.1             0.000000
case_number.2             0.000000
original_order            0.000000
unnamed:_22               0.999833
unnamed:_23               0.999666
dtype: float64

I checked specifically which columns have more than 70% missing values

In [23]:
missing_values_perc.loc[missing_values_perc.gt(0.7)].index

Index(['unnamed:_22', 'unnamed:_23'], dtype='object')

So I removed these columns: 'Unnamed: 22', 'Unnamed: 23' because 99.9% missing values

In [24]:
sharks_copy = sharks_copy.drop(columns=['unnamed:_22', 'unnamed:_23'])

Then I double checked that it worked by checking the shape and comparing that to before I made the change. 
Previous output: (5992, 24). Current output: (5992, 22)

In [25]:
sharks_copy.shape

(5992, 22)

## Check for duplicate columns

A few columns looked very similar, so I checked if they were copies of each other.
I found that they were not, so they could not be deleted without further inspection. I decided to leave them as they are and rather work on cleaning the data in other ways.

In [26]:
'Case_Number.1' == 'Case_Number.2'

False

In [27]:
'Case_Number.1' == 'Case_Number'

False

In [28]:
'Case_Number.2' == 'Case_Number'

False

In [29]:
'href_formula' == 'href'

False

## Check for incorrectly named values in a column & rename them

I saw that there are a few unexpected values in the column 'sex', so I investigated further and renamed them as appropriate

In [30]:
sharks_copy['sex'].value_counts()

M      4835
F       585
M         2
lli       1
.         1
N         1
Name: sex, dtype: int64

In [31]:
print(set(sharks_copy['sex']))

{nan, 'M', 'N', '.', 'M ', 'F', 'lli'}


In [32]:
sharks_copy['sex'].unique()

array(['M', nan, 'F', 'M ', 'lli', 'N', '.'], dtype=object)

'M ' clearly meant 'M', so I replaced it, so that these values are correctly counted as male

In [33]:
sharks_copy['sex'] = sharks_copy['sex'].str.replace('M ', 'M')

In [34]:
sharks_copy['sex'].value_counts()

M      4837
F       585
lli       1
.         1
N         1
Name: sex, dtype: int64

In [35]:
print(set(sharks_copy['sex']))

{nan, 'M', 'N', '.', 'F', 'lli'}


I checked which row has 'lli' in it. Name is 'Brian Kang' --> most likely male, so I was able to change 'lli' to 'M' as well

In [36]:
sharks_copy.loc[sharks_copy['sex'] == 'lli']

case_number       date  year incident_type country        area  \
1400  2004.11.11.b  11-Nov-04  2004    Unprovoked     USA  California   

                                            location activity        name  \
1400  Bunkers, Humboldt Bay, Eureka, Humboldt County  Surfing  Brian Kang   

      sex  ... fatal_(y/n)   time                  species  \
1400  lli  ...           N  13h30  5.5 m [18'] white shark   

                investigator_or_source                    pdf  \
1400  R. Collier, GSAF                  2004.11.11.b-Kang.pdf   

                                           href_formula  \
1400  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
1400  http://sharkattackfile.net/spreadsheets/pdf_di...  2004.11.11.b   

     case_number.2 original_order  
1400  2004.11.11.b           4593  

[1 rows x 22 columns]

In [37]:
sharks_copy['sex'] = sharks_copy['sex'].str.replace('lli', 'M')

I checked the row with the value 'N' in it. 
I expected that it might refer to a third gender, but on closer inspection it doesn't appear so.
From the names and year, this appears to be a married heterosexual couple.
Therefore, the value 'N' for 'sex' is misleading and should be renamed. 

In [38]:
sharks_copy.loc[sharks_copy['sex'] == 'N']

case_number       date  year incident_type    country             area  \
4708  1934.07.11  11-Jul-34  1934       Boating  AUSTRALIA  New South Wales   

      location activity                                         name sex  ...  \
4708  Cronulla  Fishing  18' boat, occupants William & Leslie Newton   N  ...   

     fatal_(y/n) time             species  \
4708           N  NaN  Blue pointer, 11'    

                                 investigator_or_source  \
4708  G.P. Whitley, ref: Daily Telegraph, 7/11/1934 ...   

                                       pdf  \
4708  1934.07.11-Newton-boat-Australia.pdf   

                                           href_formula  \
4708  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
4708  http://sharkattackfile.net/spreadsheets/pdf_di...    1934.07.11   

     case_number.2 original_order  
4708    1934.07.11           1285  

[1 rows x 22 columns]

In [39]:
sharks_copy['sex'] = sharks_copy['sex'].str.replace('N', 'Unknown')

Then I checked the last unclear value in the column, and decided that it should also be replaced with 'Unknown' as I could not discern other relevant information indicating otherwise from looking at other data in the row.

In [40]:
sharks_copy.loc[sharks_copy['sex'] == '.']

case_number                  date  year incident_type  \
5201  1908.06.02.R  Reported 02-Jun-1908  1908  Sea Disaster   

               country         area location activity name sex  ...  \
5201  PAPUA NEW GUINEA  New Britain   Matupi        .  NaN   .  ...   

     fatal_(y/n) time                    species     investigator_or_source  \
5201           Y  NaN  Allegedly a 33-foot shark  Taranaki Herald, 6/2/1908   

                          pdf  \
5201  1908.06.02.R-Matupi.pdf   

                                           href_formula  \
5201  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
5201  http://sharkattackfile.net/spreadsheets/pdf_di...  1908.06.02.R   

     case_number.2 original_order  
5201  1908.06.02.R            792  

[1 rows x 22 columns]

In [41]:
sharks_copy['sex'] = sharks_copy['sex'].str.replace('.', 'Unknown')

In [42]:
sharks_copy['sex'].value_counts()

M          4838
F           585
Unknown       2
Name: sex, dtype: int64

## Changing values in a column depending on a condition

I saw that in the 'name' column, there were several values with 'male' that had a null value under 'sex'.

In [43]:
sharks_copy['sex'].value_counts()

M          4838
F           585
Unknown       2
Name: sex, dtype: int64

In [44]:
sharks_copy.loc[sharks_copy['name'] == 'male']

case_number         date  year incident_type             country  \
0     2016.09.18.c    18-Sep-16  2016    Unprovoked                 USA   
2     2016.09.18.a    18-Sep-16  2016    Unprovoked                 USA   
4       2016.09.15    16-Sep-16  2016    Unprovoked           AUSTRALIA   
6       2016.09.11    11-Sep-16  2016    Unprovoked                 USA   
14    2016.08.29.a    29-Aug-16  2016    Unprovoked                 USA   
...            ...          ...   ...           ...                 ...   
5963       ND.0030  Before 1905     0    Unprovoked               BURMA   
5975       ND.0017  Before 1921     0    Unprovoked        SOUTH AFRICA   
5976       ND.0016  Before 1921     0    Unprovoked        SOUTH AFRICA   
5987       ND.0005  Before 1903     0    Unprovoked           AUSTRALIA   
5991       ND.0001    1845-1853     0    Unprovoked  CEYLON (SRI LANKA)   

                   area                             location  \
0               Florida     New Smyrna Beach, Volusia County   
2               Florida     New Smyrna Beach, Volusia County   
4              Victoria                          Bells Beach   
6               Florida        Ponte Vedra, St. Johns County   
14              Florida     New Smyrna Beach, Volusia County   
...                 ...                                  ...   
5963                NaN                                  NaN   
5975      KwaZulu-Natal                              Durban    
5976      KwaZulu-Natal                               Durban   
5987  Western Australia                          Roebuck Bay   
5991   Eastern Province  Below the English fort, Trincomalee   

                                         activity  name sex  ... fatal_(y/n)  \
0                                         Surfing  male   M  ...           N   
2                                         Surfing  male   M  ...           N   
4                                         Surfing  male   M  ...           N   
6                                          Wading  male   M  ...           N   
14                                        Surfing  male   M  ...           N   
...                                           ...   ...  ..  ...         ...   
5963                                      Bathing  male   M  ...           Y   
5975  Crew swimming alongside their anchored ship  male   M  ...           Y   
5976                           4 men were bathing  male   M  ...           Y   
5987                                       Diving  male   M  ...           Y   
5991                                     Swimming  male   M  ...           Y   

       time         species  \
0     13h00             NaN   
2     10h43             NaN   
4       NaN       2 m shark   
6     15h15  3' to 4' shark   
14    14h00             NaN   
...     ...             ...   
5963    NaN             NaN   
5975    NaN             NaN   
5976    NaN             NaN   
5987    NaN             NaN   
5991    NaN             NaN   

                                 investigator_or_source  \
0                           Orlando Sentinel, 9/19/2016   
2                           Orlando Sentinel, 9/19/2016   
4                                    The Age, 9/16/2016   
6                                   News4Jax, 9/11/2016   
14                              News Channel 8, 8/30/16   
...                                                 ...   
5963                      Massillon Independent, 3/1905   
5975  Captain A. Anderson,  Natal Mercury, 12/31/192...   
5976  Captain A. Anderson,  Natal Mercury, 12/31/192...   
5987                   H. Taunton; N. Bartlett,  p. 234   
5991                                         S.W. Baker   

                                 pdf  \
0               2016.09.18.c-NSB.pdf   
2               2016.09.18.a-NSB.pdf   
4          2016.09.16-BellsBeach.pdf   
6          2016.09.11-PonteVedra.pdf   
14              2016.08.29.a-NSB.pdf   
...                              ...   
5963          

This can be seen clearly in the table below

In [45]:
sharks_copy.loc[(sharks_copy['name'] == 'male') & (sharks_copy['sex'].isnull())]

case_number                   date  year incident_type       country  \
2000    1997.09.08              08-Sep-97  1997    Unprovoked           USA   
2824    1980.12.26              26-Dec-80  1980       Invalid  SOUTH AFRICA   
3068    1973.09.29              29-Sep-73  1973  Sea Disaster  SOUTH AFRICA   
4112  1954.07.01.R   Reported 01-Jul-1954  1954       Invalid       CROATIA   
4240    1950.07.19             1950.07.19  1950      Provoked         ITALY   
5018    1920.11.04              04-Nov-20  1920  Sea Disaster   PHILIPPINES   
5779    1842.07.06            06-Jul-1842  1842      Provoked           USA   
5825  1816.09.03.R  Reported 03-Sept-1816  1816    Unprovoked           USA   
5865    0500.00.00         Circa 500 A.D.   500    Unprovoked        MEXICO   

                       area                          location  \
2000                Florida  New Smyrna Beach, Volusia County   
2824  Eastern Cape Province                    Port Elizabeth   
3068          KwaZulu-Natal                     Mission Rocks   
4112                    NaN                              Pula   
4240                 Savona                           Albenga   
5018                  Leyte                               NaN   
5779             New Jersey          Absecon, Atlantic County   
5825           Rhode Island                    Bristol Harbor   
5865                    NaN                               NaN   

                                               activity  name  sex  ...  \
2000                                            Surfing  male  NaN  ...   
2824                                                NaN  male  NaN  ...   
3068  Being pulled to shore from wreck of 25-ton fis...  male  NaN  ...   
4112                                                NaN  male  NaN  ...   
4240                                            Fishing  male  NaN  ...   
5018  The coastwise steamer San Basilio capsized in ...  male  NaN  ...   
5779                                  Harassing a shark  male  NaN  ...   
5825                                           Swimming  male  NaN  ...   
5865                                                NaN  male  NaN  ...   

     fatal_(y/n)     time               species  \
2000           N  Evening  small blacktip shark   
2824           Y      NaN                   NaN   
3068           Y      NaN                   NaN   
4112         NaN      NaN                   NaN   
4240           N      NaN                   NaN   
5018           Y      NaN                   NaN   
5779           n      NaN                   NaN   
5825           Y      NaN                   NaN   
5865           N      NaN                   NaN   

                   investigator_or_source                            pdf  \
2000       Daytona News-Journal, 9/9/1997             1997.09.08-NSB.pdf   
2824  Eastern Province Herald, 12/29/1980      1980.12.26-scavenging.pdf   
3068             Natal Mercury, 10/5/1973           1973.09.29-AlanS.pdf   
4112                       C. Moore, GSAF          1954.07.01.R-Pula.pdf   
4240                       C. Moore, GSAF         1950.07.19-Albenga.pdf   
5018          Oakland Tribune, 11/11/1920     1920.11.04-Philippines.pdf   
5779     New York Evening Post, 7/11/1842         1842.07.06-Absecon.pdf   
5825        Connecticut Courant, 9/3/1816  1816.09.03.R-Rhode-Island.pdf   
5865                            J. Castro               500AD-Mexico.pdf   

                                           href_formula  \
2000  http://sharkattackfile.net/spreadsheets/pdf_di...   
2824  http://sharkattackfile.net/spreadsheets/pdf_di...   
3068  http://sharkattackfile.net/spreadsheets/pdf_di...   
4112  http://sharkattackfile.net/spreadsheets/pdf_di...   
4240  http://sharkattackfile.net/spreadsheets/pdf_di...   
5018  http://sharkattackfile.net/spreadsheets/pdf_di...   
5779  http://sharkattackfile.net/spreadsheets/pdf_di...   
5825  http://sharkattackfile.net/spreadsheets/pdf_di...   
5865  http:

I set a condition to change the null values in the 'sex' column to 'M' when 'male' fell under 'name'

In [46]:
sharks_copy.loc[(sharks_copy['name'] == 'male') & (sharks_copy['sex'].isna()), 'sex'] = 'M'

Below shows that it worked- there are now 9 more values under 'M' in the 'sex' column and recalling the table that meets the condition comes up empty. 

In [47]:
sharks_copy['sex'].value_counts()

M          4847
F           585
Unknown       2
Name: sex, dtype: int64

In [48]:
sharks_copy.loc[(sharks_copy['name'] == 'male') & (sharks_copy['sex'].isnull())]

Empty DataFrame
Columns: [case_number, date, year, incident_type, country, area, location, activity, name, sex, age, injury, fatal_(y/n), time, species, investigator_or_source, pdf, href_formula, href, case_number.1, case_number.2, original_order]
Index: []

[0 rows x 22 columns]

In [49]:
sharks_copy.loc[sharks_copy['name'] == 'male']

case_number         date  year incident_type             country  \
0     2016.09.18.c    18-Sep-16  2016    Unprovoked                 USA   
2     2016.09.18.a    18-Sep-16  2016    Unprovoked                 USA   
4       2016.09.15    16-Sep-16  2016    Unprovoked           AUSTRALIA   
6       2016.09.11    11-Sep-16  2016    Unprovoked                 USA   
14    2016.08.29.a    29-Aug-16  2016    Unprovoked                 USA   
...            ...          ...   ...           ...                 ...   
5963       ND.0030  Before 1905     0    Unprovoked               BURMA   
5975       ND.0017  Before 1921     0    Unprovoked        SOUTH AFRICA   
5976       ND.0016  Before 1921     0    Unprovoked        SOUTH AFRICA   
5987       ND.0005  Before 1903     0    Unprovoked           AUSTRALIA   
5991       ND.0001    1845-1853     0    Unprovoked  CEYLON (SRI LANKA)   

                   area                             location  \
0               Florida     New Smyrna Beach, Volusia County   
2               Florida     New Smyrna Beach, Volusia County   
4              Victoria                          Bells Beach   
6               Florida        Ponte Vedra, St. Johns County   
14              Florida     New Smyrna Beach, Volusia County   
...                 ...                                  ...   
5963                NaN                                  NaN   
5975      KwaZulu-Natal                              Durban    
5976      KwaZulu-Natal                               Durban   
5987  Western Australia                          Roebuck Bay   
5991   Eastern Province  Below the English fort, Trincomalee   

                                         activity  name sex  ... fatal_(y/n)  \
0                                         Surfing  male   M  ...           N   
2                                         Surfing  male   M  ...           N   
4                                         Surfing  male   M  ...           N   
6                                          Wading  male   M  ...           N   
14                                        Surfing  male   M  ...           N   
...                                           ...   ...  ..  ...         ...   
5963                                      Bathing  male   M  ...           Y   
5975  Crew swimming alongside their anchored ship  male   M  ...           Y   
5976                           4 men were bathing  male   M  ...           Y   
5987                                       Diving  male   M  ...           Y   
5991                                     Swimming  male   M  ...           Y   

       time         species  \
0     13h00             NaN   
2     10h43             NaN   
4       NaN       2 m shark   
6     15h15  3' to 4' shark   
14    14h00             NaN   
...     ...             ...   
5963    NaN             NaN   
5975    NaN             NaN   
5976    NaN             NaN   
5987    NaN             NaN   
5991    NaN             NaN   

                                 investigator_or_source  \
0                           Orlando Sentinel, 9/19/2016   
2                           Orlando Sentinel, 9/19/2016   
4                                    The Age, 9/16/2016   
6                                   News4Jax, 9/11/2016   
14                              News Channel 8, 8/30/16   
...                                                 ...   
5963                      Massillon Independent, 3/1905   
5975  Captain A. Anderson,  Natal Mercury, 12/31/192...   
5976  Captain A. Anderson,  Natal Mercury, 12/31/192...   
5987                   H. Taunton; N. Bartlett,  p. 234   
5991                                         S.W. Baker   

                                 pdf  \
0               2016.09.18.c-NSB.pdf   
2               2016.09.18.a-NSB.pdf   
4          2016.09.16-BellsBeach.pdf   
6          2016.09.11-PonteVedra.pdf   
14              2016.08.29.a-NSB.pdf   
...                              ...   
5963          

## Changing index

I wanted to check if the another column (original_order, case_number.1 or case_number.2) could work as an index column, but I saw that their values are not all unique, so they wouldn't work. I decided that it would be clearer to leave the index numbers as they are. 

In [50]:
sharks_copy['original_order'].value_counts(dropna=False)

5661    2
569     2
3847    2
5739    2
2047    1
       ..
5428    1
1334    1
3383    1
5432    1
2049    1
Name: original_order, Length: 5988, dtype: int64

In [51]:
sharks_copy['case_number.2'].value_counts(dropna=False)

1923.00.00.a    2
1983.06.15      2
1966.12.26      2
2012.09.02.b    2
1990.05.10      2
               ..
2008.12.20      1
1995.08.15      1
1965.03.14      1
2011.08.16.b    1
2016.05.22      1
Name: case_number.2, Length: 5976, dtype: int64

In [52]:
sharks_copy['case_number.1'].value_counts(dropna=False)

1923.00.00.a    2
2009.12.18      2
2012.09.02.b    2
1952.08.04      2
1966.12.26      2
               ..
1961.01.03.a    1
1901.01.30      1
2008.12.20      1
1995.08.15      1
2016.05.22      1
Name: case_number.1, Length: 5975, dtype: int64

## Create a new column with year categories

I thought it would be interesting to see the years grouped and be able to sort and filter the data according to a broader time period, so I decided to create an extra column with specific time periods.

STEP 1: redefine 0 values in 'year' column as 'Unknown'.
- To ensure that they do not get incorrectly counted in the 'before 1900' category. 
- Ideally, I would have liked to take a closer look at the 'date' column and reassign some of the 'Unknowns' with information provided in the 'date' column. However, given that the 'date' column is extremely messy, with various combinations of strings and integers and some very vague values (e.g. 'before 2016', 'before the war', 'ca. 336 B.C'), I decided that that would be too ambitious to look in so much detail for this short assignment, given the time constraints.  
- I looked more closely at the data and the values '5', '77' and '500', which first appeared to be incorrect values, as referred to as 'A.D' in the 'date' column, indicating that they do, in fact, refer to years. For example, the value '5' is based on an Aboriginal rock painting depicting a man being attacked by a shark.

STEP 2: create new empty column: 'year_categories'

STEP 3: set categories by decade and apply to new column

In [53]:
print(set(sharks_copy['year']))

{0, 5, 1543, 1554, 1555, 1580, 1595, 77, 1617, 1637, 1638, 1642, 1700, 1703, 1721, 1733, 1738, 1742, 1748, 1749, 1755, 1758, 1764, 1767, 1771, 1776, 1779, 1785, 1787, 1788, 1791, 1800, 1803, 1804, 1805, 1807, 1811, 1812, 1816, 1817, 1818, 1819, 1822, 1825, 1826, 1827, 1828, 1829, 1830, 1831, 1832, 1834, 1835, 1836, 1837, 1839, 1840, 1841, 1842, 1844, 1845, 1846, 1847, 1848, 1849, 1850, 1851, 1852, 1853, 1854, 1855, 1856, 1858, 1859, 1860, 1861, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1869, 1870, 1871, 1872, 1873, 1874, 1875, 1876, 1877, 1878, 1879, 1880, 1881, 1882, 1883, 1884, 1885, 1886, 1887, 1888, 1889, 1890, 1891, 1892, 1893, 1894, 1895, 1896, 1897, 1898, 1899, 1900, 1901, 1902, 1903, 1904, 1905, 1906, 1907, 1908, 1909, 1910, 1911, 1912, 1913, 1914, 1915, 1916, 1917, 1918, 1919, 1920, 1921, 1922, 1923, 1924, 1925, 1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936, 1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950, 1951, 1952, 1953,

In [54]:
sharks_copy['year'].dtypes

dtype('int64')

### Step 1: change null values to 'Unknown'

In [55]:
sharks_copy.year = sharks_copy['year'].replace([0],'Unknown')

Check that it worked

In [56]:
sharks_copy.tail()

case_number         date     year incident_type             country  \
5987     ND.0005  Before 1903  Unknown    Unprovoked           AUSTRALIA   
5988     ND.0004  Before 1903  Unknown    Unprovoked           AUSTRALIA   
5989     ND.0003    1900-1905  Unknown    Unprovoked                 USA   
5990     ND.0002    1883-1889  Unknown    Unprovoked              PANAMA   
5991     ND.0001    1845-1853  Unknown    Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
5987  Western Australia                          Roebuck Bay        Diving   
5988  Western Australia                                  NaN  Pearl diving   
5989     North Carolina                       Ocracoke Inlet      Swimming   
5990                NaN                 Panama Bay 8ºN, 79ºW           NaN   
5991   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex  ... fatal_(y/n) time species  \
5987                   male   M  ...           Y  NaN     NaN   
5988                  Ahmun   M  ...           Y  NaN     NaN   
5989  Coast Guard personnel   M  ...           Y  NaN     NaN   
5990        Jules Patterson   M  ...           Y  NaN     NaN   
5991                   male   M  ...           Y  NaN     NaN   

                     investigator_or_source                             pdf  \
5987       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
5988  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
5989   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
5990                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
5991                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href_formula  \
5987  http://sharkattackfile.net/spreadsheets/pdf_di...   
5988  http://sharkattackfile.net/spreadsheets/pdf_di...   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...   
5991  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
5987  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0005   
5988  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0004   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0003   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0002   
5991  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0001   

     case_number.2 original_order  
5987       ND.0005              6  
5988       ND.0004              5  
5989       ND.0003              4  
5990       ND.0002              3  
5991       ND.0001              2  

[5 rows x 22 columns]

In [57]:
print(set(sharks_copy['year']))

{5, 1543, 1554, 1555, 1580, 1595, 'Unknown', 77, 1617, 1637, 1638, 1642, 1700, 1703, 1721, 1733, 1738, 1742, 1748, 1749, 1755, 1758, 1764, 1767, 1771, 1776, 1779, 1785, 1787, 1788, 1791, 1800, 1803, 1804, 1805, 1807, 1811, 1812, 1816, 1817, 1818, 1819, 1822, 1825, 1826, 1827, 1828, 1829, 1830, 1831, 1832, 1834, 1835, 1836, 1837, 1839, 1840, 1841, 1842, 1844, 1845, 1846, 1847, 1848, 1849, 1850, 1851, 1852, 1853, 1854, 1855, 1856, 1858, 1859, 1860, 1861, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1869, 1870, 1871, 1872, 1873, 1874, 1875, 1876, 1877, 1878, 1879, 1880, 1881, 1882, 1883, 1884, 1885, 1886, 1887, 1888, 1889, 1890, 1891, 1892, 1893, 1894, 1895, 1896, 1897, 1898, 1899, 1900, 1901, 1902, 1903, 1904, 1905, 1906, 1907, 1908, 1909, 1910, 1911, 1912, 1913, 1914, 1915, 1916, 1917, 1918, 1919, 1920, 1921, 1922, 1923, 1924, 1925, 1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936, 1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950, 1951, 195

### Step 2: create a new empty column

In [58]:
sharks_copy = sharks_copy.assign(year_categories=pd.NA)

Check that it worked

In [59]:
print(set(sharks_copy['year_categories']))

{<NA>}


In [60]:
sharks_copy['year'].dtypes

dtype('O')

I encountered an issue when I tried to run the below - I was unable to assign values to my new year_categories column. It looks like, in changing the null values to 'Unknown', I also changed the data type of my column. Therefore, I had to change the data type of the 'year' column back to int.

In [61]:
sharks_copy.year = pd.to_numeric(sharks_copy.year, errors='coerce', downcast='integer')

In [62]:
sharks_copy['year'] = sharks_copy['year'].astype('Int64')

In [63]:
sharks_copy['year'].dtypes

Int64Dtype()

### Step 3: create categories and assign values to them in the new column

After doing the above, I was able to assign values to my new column as I had planned to.

In [64]:
sharks_copy.loc[sharks_copy["year"].le(1899), "year_categories"] = "before 1900"

In [65]:
sharks_copy.loc[sharks_copy["year"].between(1900, 1949), "year_categories"] = "1900 to 1949"

In [66]:
sharks_copy.loc[sharks_copy["year"].between(1950, 1999), "year_categories"] = "1950 to 1999"

In [67]:
sharks_copy.loc[sharks_copy["year"].ge(2000), "year_categories"] = "from 2000 onwards"

Check that it worked:

In [68]:
sharks_copy.head()

case_number       date  year incident_type    country      area  \
0  2016.09.18.c  18-Sep-16  2016    Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016    Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016    Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   

                           location activity             name sex  ...   time  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  ...  13h00   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  ...  11h00   
2  New Smyrna Beach, Volusia County  Surfing             male   M  ...  10h43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  ...    NaN   
4                       Bells Beach  Surfing             male   M  ...    NaN   

     species       investigator_or_source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.16   

  case_number.2 original_order    year_categories  
0  2016.09.18.c           5993  from 2000 onwards  
1  2016.09.18.b           5992  from 2000 onwards  
2  2016.09.18.a           5991  from 2000 onwards  
3    2016.09.17           5990  from 2000 onwards  
4    2016.09.15           5989  from 2000 onwards  

[5 rows x 23 columns]

In [69]:
sharks_copy.tail()

case_number         date  year incident_type             country  \
5987     ND.0005  Before 1903  <NA>    Unprovoked           AUSTRALIA   
5988     ND.0004  Before 1903  <NA>    Unprovoked           AUSTRALIA   
5989     ND.0003    1900-1905  <NA>    Unprovoked                 USA   
5990     ND.0002    1883-1889  <NA>    Unprovoked              PANAMA   
5991     ND.0001    1845-1853  <NA>    Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
5987  Western Australia                          Roebuck Bay        Diving   
5988  Western Australia                                  NaN  Pearl diving   
5989     North Carolina                       Ocracoke Inlet      Swimming   
5990                NaN                 Panama Bay 8ºN, 79ºW           NaN   
5991   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex  ... time species  \
5987                   male   M  ...  NaN     NaN   
5988                  Ahmun   M  ...  NaN     NaN   
5989  Coast Guard personnel   M  ...  NaN     NaN   
5990        Jules Patterson   M  ...  NaN     NaN   
5991                   male   M  ...  NaN     NaN   

                     investigator_or_source                             pdf  \
5987       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
5988  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
5989   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
5990                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
5991                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href_formula  \
5987  http://sharkattackfile.net/spreadsheets/pdf_di...   
5988  http://sharkattackfile.net/spreadsheets/pdf_di...   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...   
5991  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
5987  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0005   
5988  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0004   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0003   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0002   
5991  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0001   

     case_number.2 original_order year_categories  
5987       ND.0005              6            <NA>  
5988       ND.0004              5            <NA>  
5989       ND.0003              4            <NA>  
5990       ND.0002              3            <NA>  
5991       ND.0001              2            <NA>  

[5 rows x 23 columns]

In [70]:
sharks_copy['year_categories'].value_counts()

1950 to 1999         2416
from 2000 onwards    1855
1900 to 1949         1053
before 1900           544
Name: year_categories, dtype: int64

In [71]:
print(set(sharks_copy['year_categories']))

{'1950 to 1999', 'from 2000 onwards', 'before 1900', '1900 to 1949', <NA>}


In [72]:
sharks_copy.loc[sharks_copy['year'] == 1959]

case_number       date  year incident_type           country  \
3808    1959.12.29  29-Dec-59  1959          Boat         AUSTRALIA   
3809    1959.12.28  28-Dec-59  1959          Boat         AUSTRALIA   
3810    1959.12.26  26-Dec-59  1959    Unprovoked  PAPUA NEW GUINEA   
3811  1959.12.19.b  19-Dec-59  1959    Unprovoked         AUSTRALIA   
3812  1959.12.19.a  19-Dec-59  1959  Sea Disaster       PHILIPPINES   
...            ...        ...   ...           ...               ...   
3896  1959.00.00.e       1959  1959      Provoked    ATLANTIC OCEAN   
3897  1959.00.00.d       1959  1959    Unprovoked  PAPUA NEW GUINEA   
3898  1959.00.00.c       1959  1959    Unprovoked  PAPUA NEW GUINEA   
3899  1959.00.00.b       1959  1959    Unprovoked  PAPUA NEW GUINEA   
3900  1959.00.00.a       1959  1959      Provoked  PAPUA NEW GUINEA   

                                            area  \
3808                                  Queensland   
3809                             New South Wales   
3810                            Rigo subdistrict   
3811                                  Queensland   
3812                                     Masbate   
...                                          ...   
3896                                         NaN   
3897                   West New Britain Province   
3898                   West New Britain Province   
3899  New Ireland Province, Bismarck Archipelago   
3900                             Morobe Province   

                                      location  \
3808                       Scotts Point Beach    
3809                         Leichardt, Sydney   
3810                                  Kaparoka   
3811  Off Wynnum in Moreton Bay, near Brisbane   
3812                                     Balud   
...                                        ...   
3896                                       NaN   
3897                        Poi, Kombe Talasea   
3898                           Kalapiai, Kombe   
3899                      Enuk Island, Kavieng   
3900                                       NaN   

                                             activity  \
3808                                         Paddling   
3809                                          Fishing   
3810             Swimming underneath house on pilings   
3811  Dived from dinghy to retrieve oar in heavy seas   
3812              ship M.V. Rizal sank during typhoon   
...                                               ...   
3896    Paddling & sailing from Buenos Aires to Miami   
3897                                    Spear fishing   
3898                                          Fishing   
3899                                              NaN   
3900                                          Fishing   

                                                   name  sex  ...        time  \
3808     12' ski, occupants: Bill Dyer & Cliff Burgess   NaN  ...   Afternoon   
3809  plywood dinghy, occupants: Jack Deegan & Trevo...  NaN  ...  Late night   
3810                                        Manama Mari  NaN  ...         NaN   
3811                              Stanley Arthur Mullen    M  ...       06h00   
3812               Mamerto Daanong, Tomas Inog & others    M  ...       12h15   
...                                                 ...  ...  ...         ...   
3896                                     Mirco Tapavica    M  ...         NaN   
3897                                               male    M  ...         NaN   
3898                                               male    M  ...         NaN   
3899                                        Pasinganlas    M  ...         NaN   
3900                                               male    M  ...         NaN   

                      species  \
3808  3.7 m [12'] tiger shark   
3809        2.4 m  [8'] shark   
3810                      NaN   
3811                      NaN   
3812                      NaN   
...                       ...   
3896       Tiger shark, 12' ?   
3897  

## Create another new column with decade categories

When examining the value counts for my new column (year_categories), I noticed that the more recent time periods have a lot more values than the earlier ones. This is likely because it has become easier to collect and record data. I felt that the time periods I chose were not the most useful and I wanted to improve this, so I decided to create an additional new column separating out the years after 1950 into decades. 

Step 1: create new column: 'year_categories_decades'

Step 2: set categories by decade and apply to new column

In [74]:
sharks_copy = sharks_copy.assign(year_categories_decades=pd.NA)

In [75]:
sharks_copy.loc[sharks_copy["year"].le(1949), "year_categories_decades"] = "before 1950"
sharks_copy.loc[sharks_copy["year"].between(1950, 1959), "year_categories_decades"] = '1950 to 1959'
sharks_copy.loc[sharks_copy["year"].between(1960, 1969), "year_categories_decades"] = '1960 to 1969'
sharks_copy.loc[sharks_copy["year"].between(1970, 1979), "year_categories_decades"] = '1970 to 1979'
sharks_copy.loc[sharks_copy["year"].between(1980, 1989), "year_categories_decades"] = '1980 to 1989'
sharks_copy.loc[sharks_copy["year"].between(1990, 1999), "year_categories_decades"] = '1990 to 1999'
sharks_copy.loc[sharks_copy["year"].between(2000, 2009), "year_categories_decades"] = '2000 to 2009'
sharks_copy.loc[sharks_copy["year"].ge(2010), "year_categories_decades"] = "from 2010 onwards"

In [76]:
sharks_copy['year_categories_decades'].value_counts()


before 1950          1597
2000 to 2009         1020
from 2010 onwards     835
1960 to 1969          617
1990 to 1999          568
1950 to 1959          463
1980 to 1989          434
1970 to 1979          334
Name: year_categories_decades, dtype: int64

I compared the missing values from 'year' with my two new columns, to ensure that they have the same number of missing values (124)

In [77]:
sharks_copy.isna().sum()

case_number                   0
date                          0
year                        124
incident_type                 0
country                      43
area                        402
location                    496
activity                    527
name                        200
sex                         558
age                        2681
injury                       27
fatal_(y/n)                  19
time                       3213
species                    2934
investigator_or_source       15
pdf                           0
href_formula                  1
href                          3
case_number.1                 0
case_number.2                 0
original_order                0
year_categories             124
year_categories_decades     124
dtype: int64

In [78]:
print(set(sharks_copy['year_categories_decades']))

{'1960 to 1969', 'before 1950', '1990 to 1999', 'from 2010 onwards', '1980 to 1989', '2000 to 2009', '1950 to 1959', '1970 to 1979', <NA>}


In [79]:
sharks_copy.head()

case_number       date  year incident_type    country      area  \
0  2016.09.18.c  18-Sep-16  2016    Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016    Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016    Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   

                           location activity             name sex  ...  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  ...   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  ...   
2  New Smyrna Beach, Volusia County  Surfing             male   M  ...   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  ...   
4                       Bells Beach  Surfing             male   M  ...   

     species       investigator_or_source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.16   

  case_number.2 original_order    year_categories year_categories_decades  
0  2016.09.18.c           5993  from 2000 onwards       from 2010 onwards  
1  2016.09.18.b           5992  from 2000 onwards       from 2010 onwards  
2  2016.09.18.a           5991  from 2000 onwards       from 2010 onwards  
3    2016.09.17           5990  from 2000 onwards       from 2010 onwards  
4    2016.09.15           5989  from 2000 onwards       from 2010 onwards  

[5 rows x 24 columns]

In [80]:
sharks_copy.loc[sharks_copy['year_categories_decades'] == 'before 1950']

case_number            date  year incident_type    country  \
4271  1949.12.00.b          Dec-49  1949  Sea Disaster        NaN   
4272  1949.12.00.a          Dec-49  1949    Unprovoked  AUSTRALIA   
4273    1949.11.20       20-Nov-49  1949    Unprovoked  AUSTRALIA   
4274    1949.11.12       12-Nov-49  1949       Invalid  AUSTRALIA   
4275  1949.08.28.b       28-Aug-49  1949    Unprovoked  AUSTRALIA   
...            ...             ...   ...           ...        ...   
5863    1554.00.00        Ca. 1554  1554    Unprovoked     FRANCE   
5864    1543.00.00        Ca. 1543  1543    Unprovoked  VENEZUELA   
5865    0500.00.00  Circa 500 A.D.   500    Unprovoked     MEXICO   
5866    0077.00.00        77  A.D.    77    Unprovoked        NaN   
5867    0005.00.00      Ca. 5 A.D.     5    Unprovoked  AUSTRALIA   

                             area                         location  \
4271                Caribbean Sea      Between  Cuba  & Costa Rica   
4272                     Victoria                         Seaholme   
4273              New South Wales             Kurnell, Botany Bay    
4274                     Victoria                 Port Phillip Bay   
4275                   Queensland  Yorkey’s Knob Beach near Cairns   
...                           ...                              ...   
5863            Nice & Marseilles                              NaN   
5864  Magarita or Cubagua Islands                              NaN   
5865                          NaN                              NaN   
5866                   Ionian Sea                              NaN   
5867              New South Wales                            Bondi   

                                            activity  \
4271  Sea Disaster, sinking of the motorship Wingate   
4272             Lying on the bottom of a 16' dinghy   
4273             Free diving or wading back to shore   
4274                                      No details   
4275                                        Swimming   
...                                              ...   
5863                                             NaN   
5864                                    Pearl diving   
5865                                             NaN   
5866                                   Sponge diving   
5867                                             NaN   

                                     name sex  ...  \
4271  Albert Battles, James Dean & 4 crew   M  ...   
4272                          Doug Miller   M  ...   
4273                 William Edward Brown   M  ...   
4274                        John W. Smith   M  ...   
4275                 Brian Ware (rescuer)   M  ...   
...                                   ...  ..  ...   
5863                males (wearing armor)   M  ...   
5864                         Indian slave   M  ...   
5865                                 male   M  ...   
5866                                males   M  ...   
5867                                 male   M  ...   

                               species  \
4271   Shark involvement not confirmed   
4272  Grey nurse shark, 2.6 m [8.5']     
4273                          6' shark   
4274                               NaN   
4275                               NaN   
...                                ...   
5863             Possibly white sharks   
5864                               NaN   
5865                               NaN   
5866                               NaN   
5867                               NaN   

                                 investigator_or_source  \
4271                           Canberra Times, 1/6/1950   
4272                 V.M. Coppleson (1958), pp. 181-182   
4273                  Sydney Morning Herald, 11/20/1949   
4274  G.P. Whitley (1951), p.194, cites Sunday Heral...   
4275                        V.M. Coppleson (1958), p.89   
...                                                 ...   
5863                                       G. Rondelet    
5864                                  

In [90]:
sharks = sharks_copy

In [91]:
sharks.head()

case_number       date  year incident_type    country      area  \
0  2016.09.18.c  18-Sep-16  2016    Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016    Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016    Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016    Unprovoked  AUSTRALIA  Victoria   

                           location activity             name sex  ...  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  ...   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  ...   
2  New Smyrna Beach, Volusia County  Surfing             male   M  ...   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  ...   
4                       Bells Beach  Surfing             male   M  ...   

     species       investigator_or_source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.16   

  case_number.2 original_order    year_categories year_categories_decades  
0  2016.09.18.c           5993  from 2000 onwards       from 2010 onwards  
1  2016.09.18.b           5992  from 2000 onwards       from 2010 onwards  
2  2016.09.18.a           5991  from 2000 onwards       from 2010 onwards  
3    2016.09.17           5990  from 2000 onwards       from 2010 onwards  
4    2016.09.15           5989  from 2000 onwards       from 2010 onwards  

[5 rows x 24 columns]

## Export as new CSV file

In [95]:
sharks.to_csv('sharks_clean_IR.csv', sep=',', index=False)